File: backend/tests/Unit/Entity/UserTest.php
<?php

namespace App\Tests\Unit\Entity;

use App\Entity\User;
use PHPUnit\Framework\TestCase;

class UserTest extends TestCase
{
    private User $user;

    protected function setUp(): void
    {
        $this->user = new User();
    }

    public function testDefaultStatusIsPending(): void
    {
        $this->assertSame('PENDING', $this->user->getStatus());
    }

    public function testDefaultIsVerifiedFalse(): void
    {
        $this->assertFalse($this->user->isVerified());
    }

    public function testCreatedAtIsSetOnConstruct(): void
    {
        $this->assertNotNull($this->user->getCreatedAt());
    }

    public function testGetRolesAlwaysIncludesRoleUser(): void
    {
        $this->user->setRoles([]);
        $this->assertContains('ROLE_USER', $this->user->getRoles());
    }

    public function testGetRolesMergesCustomRoleWithRoleUser(): void
    {
        $this->user->setRoles(['ROLE_ADMIN']);
        $roles = $this->user->getRoles();
        $this->assertContains('ROLE_USER', $roles);
        $this->assertContains('ROLE_ADMIN', $roles);
    }

    public function testGetRolesDeduplicatesRoleUser(): void
    {
        $this->user->setRoles(['ROLE_USER', 'ROLE_USER']);
        $roles = $this->user->getRoles();
        $this->assertCount(1, array_filter($roles, fn($r) => $r === 'ROLE_USER'));
    }

    public function testGetPrimaryRoleReturnsNonUserRole(): void
    {
        $this->user->setRoles(['ROLE_THERAPIST']);
        $this->assertSame('ROLE_THERAPIST', $this->user->getPrimaryRole());
    }

    public function testGetPrimaryRoleDefaultsToRoleUserWhenNoneSet(): void
    {
        $this->user->setRoles([]);
        $this->assertSame('ROLE_USER', $this->user->getPrimaryRole());
    }

    public function testGetFullNameCombinesFirstAndLastName(): void
    {
        $this->user->setFirstName('John');
        $this->user->setLastName('Doe');
        $this->assertSame('John Doe', $this->user->getFullName());
    }

    public function testGetFullNameWithNullFirstName(): void
    {
        $this->user->setFirstName(null);
        $this->user->setLastName('Doe');
        $this->assertSame('Doe', $this->user->getFullName());
    }

    public function testGetFullNameWithNullLastName(): void
    {
        $this->user->setFirstName('John');
        $this->user->setLastName(null);
        $this->assertSame('John', $this->user->getFullName());
    }

    public function testGetFullNameReturnEmptyStringWhenBothNull(): void
    {
        $this->user->setFirstName(null);
        $this->user->setLastName(null);
        $this->assertSame('', $this->user->getFullName());
    }

    public function testGetProfilePictureUrlReturnsNullWhenNoProfilePicture(): void
    {
        $this->assertNull($this->user->getProfilePictureUrl());
    }

    public function testGetProfilePictureUrlReturnsHttpUrlUnchanged(): void
    {
        $this->user->setProfilePicture('https://example.com/photo.jpg');
        $this->assertSame('https://example.com/photo.jpg', $this->user->getProfilePictureUrl());
    }

    public function testGetProfilePictureUrlReturnsRelativeUploadPathUnchanged(): void
    {
        $this->user->setProfilePicture('/uploads/profiles/photo.jpg');
        $this->assertSame('/uploads/profiles/photo.jpg', $this->user->getProfilePictureUrl());
    }

    public function testGetProfilePictureUrlExtractsFilenameFromAbsoluteWindowsPath(): void
    {
        $this->user->setProfilePicture('C:\\Users\\user\\Pictures\\photo.jpg');
        $this->assertSame('/uploads/profiles/photo.jpg', $this->user->getProfilePictureUrl());
    }

    public function testGetUserIdentifierReturnsEmail(): void
    {
        $this->user->setEmail('user@test.com');
        $this->assertSame('user@test.com', $this->user->getUserIdentifier());
    }

    public function testSetEmailReturnsSelf(): void
    {
        $result = $this->user->setEmail('test@example.com');
        $this->assertSame($this->user, $result);
    }

    public function testSetIsVerifiedReturnsSelf(): void
    {
        $result = $this->user->setIsVerified(true);
        $this->assertSame($this->user, $result);
        $this->assertTrue($this->user->isVerified());
    }

    public function testCodeCollectionIsInitializedOnConstruct(): void
    {
        $this->assertCount(0, $this->user->getCode());
    }
}

File: frontend/src/test/java/com/innertrack/model/UserTest.java
package com.innertrack.model;

import org.junit.jupiter.api.BeforeEach;
import org.junit.jupiter.api.Test;

import java.time.LocalDateTime;
import java.util.Arrays;
import java.util.List;

import static org.junit.jupiter.api.Assertions.*;

class UserTest {

    private User user;

    @BeforeEach
    void setUp() {
        user = new User();
    }

    @Test
    void testGetFullNameWithBothNames() {
        user.setFirstName("John");
        user.setLastName("Doe");
        assertEquals("John Doe", user.getFullName());
    }

    @Test
    void testGetFullNameWithNullFirstName() {
        user.setFirstName(null);
        user.setLastName("Doe");
        assertEquals("Doe", user.getFullName());
    }

    @Test
    void testGetFullNameWithNullLastName() {
        user.setFirstName("John");
        user.setLastName(null);
        assertEquals("John", user.getFullName());
    }

    @Test
    void testGetFullNameWithBothNullReturnsEmptyString() {
        user.setFirstName(null);
        user.setLastName(null);
        assertEquals("", user.getFullName());
    }

    @Test
    void testGetFullNameWithEmptyStringsReturnsEmptyString() {
        user.setFirstName("");
        user.setLastName("");
        assertEquals("", user.getFullName());
    }

    @Test
    void testGetFullNameTrimsTrailingSpaces() {
        user.setFirstName("John");
        user.setLastName(null);
        assertFalse(user.getFullName().endsWith(" "));
    }

    @Test
    void testIsVerifiedDefaultsFalse() {
        assertFalse(user.isVerified());
    }

    @Test
    void testSetVerifiedToTrue() {
        user.setVerified(true);
        assertTrue(user.isVerified());
    }

    @Test
    void testSetAndGetId() {
        user.setId(42);
        assertEquals(42, user.getId());
    }

    @Test
    void testSetAndGetEmail() {
        user.setEmail("john@example.com");
        assertEquals("john@example.com", user.getEmail());
    }

    @Test
    void testSetAndGetPassword() {
        user.setPassword("hashed_password");
        assertEquals("hashed_password", user.getPassword());
    }

    @Test
    void testSetAndGetStatus() {
        user.setStatus("ACTIVE");
        assertEquals("ACTIVE", user.getStatus());
    }

    @Test
    void testSetAndGetPhoneNumber() {
        user.setPhoneNumber("+216 12 345 678");
        assertEquals("+216 12 345 678", user.getPhoneNumber());
    }

    @Test
    void testSetAndGetProfilePicture() {
        user.setProfilePicture("/uploads/profiles/photo.jpg");
        assertEquals("/uploads/profiles/photo.jpg", user.getProfilePicture());
    }

    @Test
    void testRolesDefaultToEmptyList() {
        assertNotNull(user.getRoles());
        assertTrue(user.getRoles().isEmpty());
    }

    @Test
    void testSetAndGetRoles() {
        List<String> roles = Arrays.asList("ROLE_ADMIN", "ROLE_USER");
        user.setRoles(roles);
        assertEquals(2, user.getRoles().size());
        assertTrue(user.getRoles().contains("ROLE_ADMIN"));
        assertTrue(user.getRoles().contains("ROLE_USER"));
    }

    @Test
    void testSetAndGetCreatedAt() {
        LocalDateTime now = LocalDateTime.now();
        user.setCreatedAt(now);
        assertEquals(now, user.getCreatedAt());
    }

    @Test
    void testSetAndGetLastLogin() {
        LocalDateTime loginTime = LocalDateTime.of(2026, 4, 30, 10, 0);
        user.setLastLogin(loginTime);
        assertEquals(loginTime, user.getLastLogin());
    }

    @Test
    void testCreatedAtDefaultsToNull() {
        assertNull(user.getCreatedAt());
    }

    @Test
    void testLastLoginDefaultsToNull() {
        assertNull(user.getLastLogin());
    }
}
